README.md
<a href="https://www.buymeacoffee.com/rodrigobastosv" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" style="height: 60px !important;width: 217px !important;" ></a>

#  Introduction

This package was created to make it easy to test widgets on different sizes of devices. It provides a simple API where you can decide what height and screen width you want the tests to run on. You also have the option to choose a specific device to test your widgets on. The screen size and pixel density will suit the chosen device

# Basic Usage - SizedWidget

The most simple usage is the case you just want to pass height and width and want to run your test on.
To do that, just wrap the widget under testing in SizedWidget.

```dart

await tester.pumpWidget(
    MaterialApp(
        home: SizedWidget(
            widget: WidgetUnderTest(),
            height: 1200,
            width: 800,
        ),
    ),
);

```

This simple step will ensure your widget runs on a screen of 1200 height and 800 width. Of course you can pass any values you want
for both height and width.

# Basic Usage - DeviceWidget

Other usage is the case you want your widgets to run on the settings of a specific device. This package has a lot of device options to choose.

If you want to use this option, just wrap the widget under testing in the widget called DeviceWidget. In this case you have
to pass the desired Device. This package provides a list of diferent devices to choose.

```dart
await tester.pumpWidget(
    MaterialApp(
        home: DeviceWidget(
            widget: WidgetUnderTest(),
            device: Device.iPhone12ProMax,
        ),
    ),
);
```

## Todo

- [x] Tests

## Suggestions & Bugs

For any suggestions or bug report please head to [issue tracker][tracker].

[tracker]: https://github.com/rodrigobastosv/test_widgets_with_size/issues


CHANGELOG.md
## 1.0.1+4

* Improving documentation

## 1.0.1+3

* Adding documentation

## 1.0.1+2

* Adjusting README

## 1.0.1+1

* Exposing Devices enum on the barel file

## 1.0.1

* Adding possibility to pass own text scale factor

## 1.0.0+1

* Formatting files

## 1.0.0

* First functional release


test/src/utils_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:test_widgets_with_size/src/device.dart';
import 'package:test_widgets_with_size/src/utils.dart';

void main() {
  test(
    'getDevicePixelRatio',
    () {
      expect(getDevicePixelRatio(Device.iPhone12ProMax), 3);
      expect(getDevicePixelRatio(Device.iPhone12Pro), 3);
      expect(getDevicePixelRatio(Device.iPhone12), 3);
      expect(getDevicePixelRatio(Device.iPhone12Mini), 3);
      expect(getDevicePixelRatio(Device.iPhoneSE), 2);
      expect(getDevicePixelRatio(Device.iPhone11ProMax), 3);
      expect(getDevicePixelRatio(Device.iPhone11XSMax), 3);
      expect(getDevicePixelRatio(Device.iPhone11XR), 3);
      expect(getDevicePixelRatio(Device.iPhone11), 3);
      expect(getDevicePixelRatio(Device.iPhoneX), 3);
      expect(getDevicePixelRatio(Device.iPhoneXS), 3);
      expect(getDevicePixelRatio(Device.iPhone11Pro), 3);
      expect(getDevicePixelRatio(Device.iPhone8), 2);
      expect(getDevicePixelRatio(Device.iPhone7), 2);
      expect(getDevicePixelRatio(Device.iPhone8Plus), 3);
      expect(getDevicePixelRatio(Device.iPhone7Plus), 3);
      expect(getDevicePixelRatio(Device.iPhone6SPlus), 3);
      expect(getDevicePixelRatio(Device.iPhoneSE2), 2);
      expect(getDevicePixelRatio(Device.iPhone6S), 2);
      expect(getDevicePixelRatio(Device.iPhone6), 2);
      expect(getDevicePixelRatio(Device.iPodTouch), 2);
      expect(getDevicePixelRatio(Device.lgG5), 4);
      expect(getDevicePixelRatio(Device.lgG4), 4);
      expect(getDevicePixelRatio(Device.lgG3), 4);
      expect(getDevicePixelRatio(Device.lgOptimusG), 2);
      expect(getDevicePixelRatio(Device.lgNexus5), 3);
      expect(getDevicePixelRatio(Device.lgNexus4), 2);
      expect(getDevicePixelRatio(Device.samsungGalaxyS9Plus), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS9), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS8Plus), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS8), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS7Edge), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS7), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS6), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyS5), 3);
      expect(getDevicePixelRatio(Device.samsungGalaxyS4), 3);
      expect(getDevicePixelRatio(Device.samsungGalaxyNexus), 2);
      expect(getDevicePixelRatio(Device.samsungGalaxyNote8), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyNote4), 4);
      expect(getDevicePixelRatio(Device.samsungGalaxyNote3), 3);
      expect(getDevicePixelRatio(Device.samsungGalaxyNote2), 2);
      expect(getDevicePixelRatio(Device.samsungGalaxyNote), 2);
      expect(getDevicePixelRatio(Device.microsoftLumia1520), 2.5);
      expect(getDevicePixelRatio(Device.microsoftLumia1020), 2.4);
      expect(getDevicePixelRatio(Device.microsoftLumia925), 2.4);
      expect(getDevicePixelRatio(Device.microsoftLumia920), 2.4);
      expect(getDevicePixelRatio(Device.microsoftLumia830), 2);
      expect(getDevicePixelRatio(Device.microsoftLumia620), 1.5);
      expect(getDevicePixelRatio(Device.motorolaNexus6), 3.5);
      expect(getDevicePixelRatio(Device.htcOne), 3);
      expect(getDevicePixelRatio(Device.htc8X), 3);
      expect(getDevicePixelRatio(Device.htcEvo3D), 1.5);
      expect(getDevicePixelRatio(Device.sonyXperiaZ3), 2);
      expect(getDevicePixelRatio(Device.sonyXperiaZ), 2);
      expect(getDevicePixelRatio(Device.sonyXperiaS), 2);
      expect(getDevicePixelRatio(Device.sonyXperiaP), 1.5);
      expect(getDevicePixelRatio(Device.xyaomiRedmiNote8T), 2.75);
      expect(getDevicePixelRatio(Device.xyaomiRedmiNote6), 2.75);
      expect(getDevicePixelRatio(Device.xyaomiRedmiNote5), 2.75);
      expect(getDevicePixelRatio(Device.xyaomiMi4), 3);
      expect(getDevicePixelRatio(Device.xyaomiMi3), 3);
      expect(getDevicePixelRatio(Device.lenovoK900), 3);
    },
  );

  test(
    'getDeviceSize',
    () {
      expect(getDeviceSize(Device.iPhone12ProMax), const Size(428, 926));
      expect(getDeviceSize(Device.iPhone12Pro), const Size(390, 844));
      expect(getDeviceSize(Device.iPhone12), const Size(390, 844));
      expect(getDeviceSize(Device.iPhone12Mini), const Size(375, 812));
      expect(getDeviceSize(Device.iPhoneSE), const Size(375, 667));
      expect(getDeviceSize(Device.iPhone11ProMax), const Size(414, 896));
      expect(getDeviceSize(Device.iPhone11XSMax), const Size(414, 896));
      expect(getDeviceSize(Device.iPhone11XR), const Size(414, 896));
      expect(getDeviceSize(Device.iPhone11), const Size(414, 896));
      expect(getDeviceSize(Device.iPhoneX), const Size(375, 812));
      expect(getDeviceSize(Device.iPhoneXS), const Size(375, 812));
      expect(getDeviceSize(Device.iPhone11Pro), const Size(375, 812));
      expect(getDeviceSize(Device.iPhone8), const Size(375, 667));
      expect(getDeviceSize(Device.iPhone7), const Size(375, 667));
      expect(getDeviceSize(Device.iPhone8Plus), const Size(414, 735));
      expect(getDeviceSize(Device.iPhone7Plus), const Size(414, 735));
      expect(getDeviceSize(Device.iPhone6SPlus), const Size(414, 735));
      expect(getDeviceSize(Device.iPhoneSE2), const Size(375, 667));
      expect(getDeviceSize(Device.iPhone6S), const Size(375, 667));
      expect(getDeviceSize(Device.iPhone6), const Size(375, 667));
      expect(getDeviceSize(Device.iPodTouch), const Size(320, 568));
      expect(getDeviceSize(Device.lgG5), const Size(360, 640));
      expect(getDeviceSize(Device.lgG4), const Size(360, 640));
      expect(getDeviceSize(Device.lgG3), const Size(360, 640));
      expect(getDeviceSize(Device.lgOptimusG), const Size(384, 640));
      expect(getDeviceSize(Device.lgNexus5), const Size(360, 640));
      expect(getDeviceSize(Device.lgNexus4), const Size(384, 640));
      expect(getDeviceSize(Device.samsungGalaxyS9Plus), const Size(360, 740));
      expect(getDeviceSize(Device.samsungGalaxyS9), const Size(360, 740));
      expect(getDeviceSize(Device.samsungGalaxyS8Plus), const Size(360, 740));
      expect(getDeviceSize(Device.samsungGalaxyS8), const Size(360, 740));
      expect(getDeviceSize(Device.samsungGalaxyS7Edge), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyS7), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyS6), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyS5), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyS4), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyNexus), const Size(360, 600));
      expect(getDeviceSize(Device.samsungGalaxyNote8), const Size(360, 740));
      expect(getDeviceSize(Device.samsungGalaxyNote4), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyNote3), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyNote2), const Size(360, 640));
      expect(getDeviceSize(Device.samsungGalaxyNote), const Size(400, 640));
      expect(getDeviceSize(Device.microsoftLumia1520), const Size(432, 768));
      expect(getDeviceSize(Device.microsoftLumia1020), const Size(320, 480));
      expect(getDeviceSize(Device.microsoftLumia925), const Size(320, 480));
      expect(getDeviceSize(Device.microsoftLumia920), const Size(320, 480));
      expect(getDeviceSize(Device.microsoftLumia830), const Size(320, 480));
      expect(getDeviceSize(Device.microsoftLumia620), const Size(320, 480));
      expect(getDeviceSize(Device.motorolaNexus6), const Size(428, 926));
      expect(getDeviceSize(Device.htcOne), const Size(360, 640));
      expect(getDeviceSize(Device.htc8X), const Size(320, 480));
      expect(getDeviceSize(Device.htcEvo3D), const Size(360, 640));
      expect(getDeviceSize(Device.sonyXperiaZ3), const Size(360, 598));
      expect(getDeviceSize(Device.sonyXperiaZ), const Size(360, 640));
      expect(getDeviceSize(Device.sonyXperiaS), const Size(360, 640));
      expect(getDeviceSize(Device.sonyXperiaP), const Size(360, 640));
      expect(getDeviceSize(Device.xyaomiRedmiNote8T), const Size(393, 775));
      expect(getDeviceSize(Device.xyaomiRedmiNote6), const Size(393, 739));
      expect(getDeviceSize(Device.xyaomiRedmiNote5), const Size(393, 739));
      expect(getDeviceSize(Device.xyaomiMi4), const Size(360, 640));
      expect(getDeviceSize(Device.xyaomiMi3), const Size(360, 640));
      expect(getDeviceSize(Device.lenovoK900), const Size(360, 640));
    },
  );
}


lib/test_widgets_with_size.dart
library test_widgets_with_size;

export 'src/test_widgets_with_size.dart';
export 'src/device.dart';


lib/src/test_widgets_with_size.dart
import 'package:flutter/material.dart';
import 'package:test_widgets_with_size/src/utils.dart';

import 'device.dart';

/// Default test width. This is the same default value user by Flutter
const kTestDefaultWidth = 800.0;

/// Default test height. This is the same default value user by Flutter
const kTestDefaultHeight = 600.0;

/// Widget that make it possible to simulate any given screen size
class SizedWidget extends StatelessWidget {
  /// The only required params is the widget under testing
  const SizedWidget({
    Key? key,
    required this.widget,
    this.width = kTestDefaultWidth,
    this.height = kTestDefaultHeight,
    this.devicePixelRatio = 1.0,
    this.textScaleFactor = 1.0,
  }) : super(key: key);

  /// Widget under testing
  final Widget widget;

  /// Desired Width
  final double width;

  /// Desired Height
  final double height;

  /// Desired Pixel Ratio
  final double devicePixelRatio;

  /// Desired Text Scale Factor
  final double textScaleFactor;

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      data: MediaQueryData(
        size: Size(width, height),
        devicePixelRatio: devicePixelRatio,
        textScaleFactor: textScaleFactor,
      ),
      child: widget,
    );
  }
}

/// Widget that make it possible to simulate any given Device from the list
/// of [Device]'s
class DeviceWidget extends StatelessWidget {
  /// The only required params are the widget under testing and the desired device
  const DeviceWidget({
    Key? key,
    required this.widget,
    required this.device,
    this.textScaleFactor = 1.0,
  }) : super(key: key);

  /// Widget under testing
  final Widget widget;

  /// Device that's beeing targeted. Pick any from the [Device] enum.
  final Device device;

  /// Desired Text Scale Factor
  final double textScaleFactor;

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      data: MediaQueryData(
        size: getDeviceSize(device),
        devicePixelRatio: getDevicePixelRatio(device),
        textScaleFactor: textScaleFactor,
      ),
      child: widget,
    );
  }
}


lib/src/device.dart
/// The package comes with a list of pre defined devices.
/// This is not an exaustive list. We can add more anytime.
enum Device {
  iPhone12ProMax,
  iPhone12Pro,
  iPhone12,
  iPhone12Mini,
  iPhoneSE,
  iPhone11ProMax,
  iPhone11XSMax,
  iPhone11XR,
  iPhone11,
  iPhoneX,
  iPhoneXS,
  iPhone11Pro,
  iPhone8,
  iPhone7,
  iPhone8Plus,
  iPhone7Plus,
  iPhone6SPlus,
  iPhoneSE2,
  iPhone6S,
  iPhone6,
  iPodTouch,
  lgG5,
  lgG4,
  lgG3,
  lgOptimusG,
  lgNexus5,
  lgNexus4,
  samsungGalaxyS9Plus,
  samsungGalaxyS9,
  samsungGalaxyS8Plus,
  samsungGalaxyS8,
  samsungGalaxyS7Edge,
  samsungGalaxyS7,
  samsungGalaxyS6,
  samsungGalaxyS5,
  samsungGalaxyS4,
  samsungGalaxyNexus,
  samsungGalaxyNote8,
  samsungGalaxyNote4,
  samsungGalaxyNote3,
  samsungGalaxyNote2,
  samsungGalaxyNote,
  microsoftLumia1520,
  microsoftLumia1020,
  microsoftLumia925,
  microsoftLumia920,
  microsoftLumia830,
  microsoftLumia620,
  motorolaNexus6,
  htcOne,
  htc8X,
  htcEvo3D,
  sonyXperiaZ3,
  sonyXperiaZ,
  sonyXperiaS,
  sonyXperiaP,
  xyaomiRedmiNote8T,
  xyaomiRedmiNote6,
  xyaomiRedmiNote5,
  xyaomiMi4,
  xyaomiMi3,
  lenovoK900,
}


lib/src/utils.dart
import 'package:flutter/material.dart';

import 'device.dart';

/// Given the Device, returns its pixel ratio
double getDevicePixelRatio(Device device) {
  switch (device) {
    case Device.iPhone12ProMax:
      return 3;
    case Device.iPhone12Pro:
      return 3;
    case Device.iPhone12:
      return 3;
    case Device.iPhone12Mini:
      return 3;
    case Device.iPhoneSE:
      return 2;
    case Device.iPhone11ProMax:
      return 3;
    case Device.iPhone11XSMax:
      return 3;
    case Device.iPhone11XR:
      return 3;
    case Device.iPhone11:
      return 3;
    case Device.iPhoneX:
      return 3;
    case Device.iPhoneXS:
      return 3;
    case Device.iPhone11Pro:
      return 3;
    case Device.iPhone8:
      return 2;
    case Device.iPhone7:
      return 2;
    case Device.iPhone8Plus:
      return 3;
    case Device.iPhone7Plus:
      return 3;
    case Device.iPhone6SPlus:
      return 3;
    case Device.iPhoneSE2:
      return 2;
    case Device.iPhone6S:
      return 2;
    case Device.iPhone6:
      return 2;
    case Device.iPodTouch:
      return 2;
    case Device.lgG5:
      return 4;
    case Device.lgG4:
      return 4;
    case Device.lgG3:
      return 4;
    case Device.lgOptimusG:
      return 2;
    case Device.lgNexus5:
      return 3;
    case Device.lgNexus4:
      return 2;
    case Device.samsungGalaxyS9Plus:
      return 4;
    case Device.samsungGalaxyS9:
      return 4;
    case Device.samsungGalaxyS8Plus:
      return 4;
    case Device.samsungGalaxyS8:
      return 4;
    case Device.samsungGalaxyS7Edge:
      return 4;
    case Device.samsungGalaxyS7:
      return 4;
    case Device.samsungGalaxyS6:
      return 4;
    case Device.samsungGalaxyS5:
      return 3;
    case Device.samsungGalaxyS4:
      return 3;
    case Device.samsungGalaxyNexus:
      return 2;
    case Device.samsungGalaxyNote8:
      return 4;
    case Device.samsungGalaxyNote4:
      return 4;
    case Device.samsungGalaxyNote3:
      return 3;
    case Device.samsungGalaxyNote2:
      return 2;
    case Device.samsungGalaxyNote:
      return 2;
    case Device.microsoftLumia1520:
      return 2.5;
    case Device.microsoftLumia1020:
      return 2.4;
    case Device.microsoftLumia925:
      return 2.4;
    case Device.microsoftLumia920:
      return 2.4;
    case Device.microsoftLumia830:
      return 2;
    case Device.microsoftLumia620:
      return 1.5;
    case Device.motorolaNexus6:
      return 3.5;
    case Device.htcOne:
      return 3;
    case Device.htc8X:
      return 3;
    case Device.htcEvo3D:
      return 1.5;
    case Device.sonyXperiaZ3:
      return 2;
    case Device.sonyXperiaZ:
      return 2;
    case Device.sonyXperiaS:
      return 2;
    case Device.sonyXperiaP:
      return 1.5;
    case Device.xyaomiRedmiNote8T:
      return 2.75;
    case Device.xyaomiRedmiNote6:
      return 2.75;
    case Device.xyaomiRedmiNote5:
      return 2.75;
    case Device.xyaomiMi4:
      return 3;
    case Device.xyaomiMi3:
      return 3;
    case Device.lenovoK900:
      return 3;
  }
}

/// Given the Device, returns its size
Size getDeviceSize(Device device) {
  switch (device) {
    case Device.iPhone12ProMax:
      return const Size(428, 926);
    case Device.iPhone12Pro:
      return const Size(390, 844);
    case Device.iPhone12:
      return const Size(390, 844);
    case Device.iPhone12Mini:
      return const Size(375, 812);
    case Device.iPhoneSE:
      return const Size(375, 667);
    case Device.iPhone11ProMax:
      return const Size(414, 896);
    case Device.iPhone11XSMax:
      return const Size(414, 896);
    case Device.iPhone11XR:
      return const Size(414, 896);
    case Device.iPhone11:
      return const Size(414, 896);
    case Device.iPhoneX:
      return const Size(375, 812);
    case Device.iPhoneXS:
      return const Size(375, 812);
    case Device.iPhone11Pro:
      return const Size(375, 812);
    case Device.iPhone8:
      return const Size(375, 667);
    case Device.iPhone7:
      return const Size(375, 667);
    case Device.iPhone8Plus:
      return const Size(414, 735);
    case Device.iPhone7Plus:
      return const Size(414, 735);
    case Device.iPhone6SPlus:
      return const Size(414, 735);
    case Device.iPhoneSE2:
      return const Size(375, 667);
    case Device.iPhone6S:
      return const Size(375, 667);
    case Device.iPhone6:
      return const Size(375, 667);
    case Device.iPodTouch:
      return const Size(320, 568);
    case Device.lgG5:
      return const Size(360, 640);
    case Device.lgG4:
      return const Size(360, 640);
    case Device.lgG3:
      return const Size(360, 640);
    case Device.lgOptimusG:
      return const Size(384, 640);
    case Device.lgNexus5:
      return const Size(360, 640);
    case Device.lgNexus4:
      return const Size(384, 640);
    case Device.samsungGalaxyS9Plus:
      return const Size(360, 740);
    case Device.samsungGalaxyS9:
      return const Size(360, 740);
    case Device.samsungGalaxyS8Plus:
      return const Size(360, 740);
    case Device.samsungGalaxyS8:
      return const Size(360, 740);
    case Device.samsungGalaxyS7Edge:
      return const Size(360, 640);
    case Device.samsungGalaxyS7:
      return const Size(360, 640);
    case Device.samsungGalaxyS6:
      return const Size(360, 640);
    case Device.samsungGalaxyS5:
      return const Size(360, 640);
    case Device.samsungGalaxyS4:
      return const Size(360, 640);
    case Device.samsungGalaxyNexus:
      return const Size(360, 600);
    case Device.samsungGalaxyNote8:
      return const Size(360, 740);
    case Device.samsungGalaxyNote4:
      return const Size(360, 640);
    case Device.samsungGalaxyNote3:
      return const Size(360, 640);
    case Device.samsungGalaxyNote2:
      return const Size(360, 640);
    case Device.samsungGalaxyNote:
      return const Size(400, 640);
    case Device.microsoftLumia1520:
      return const Size(432, 768);
    case Device.microsoftLumia1020:
      return const Size(320, 480);
    case Device.microsoftLumia925:
      return const Size(320, 480);
    case Device.microsoftLumia920:
      return const Size(320, 480);
    case Device.microsoftLumia830:
      return const Size(320, 480);
    case Device.microsoftLumia620:
      return const Size(320, 480);
    case Device.motorolaNexus6:
      return const Size(428, 926);
    case Device.htcOne:
      return const Size(360, 640);
    case Device.htc8X:
      return const Size(320, 480);
    case Device.htcEvo3D:
      return const Size(360, 640);
    case Device.sonyXperiaZ3:
      return const Size(360, 598);
    case Device.sonyXperiaZ:
      return const Size(360, 640);
    case Device.sonyXperiaS:
      return const Size(360, 640);
    case Device.sonyXperiaP:
      return const Size(360, 640);
    case Device.xyaomiRedmiNote8T:
      return const Size(393, 775);
    case Device.xyaomiRedmiNote6:
      return const Size(393, 739);
    case Device.xyaomiRedmiNote5:
      return const Size(393, 739);
    case Device.xyaomiMi4:
      return const Size(360, 640);
    case Device.xyaomiMi3:
      return const Size(360, 640);
    case Device.lenovoK900:
      return const Size(360, 640);
  }
}
